packages/firestore/src/core/firestore_client.ts

```diff
@@ -199,11 +197,15 @@ export async function setOfflineComponentProvider(
   const configuration = await client.getConfiguration();
   await offlineComponentProvider.initialize(configuration);
 
-  client.setCredentialChangeListener(user =>
-    client.asyncQueue.enqueueRetryable(async () => {
-      await handleUserChange(offlineComponentProvider.localStore, user);
-    })
-  );
+  let currentUser = configuration.initialUser;
+  client.setCredentialChangeListener(user => {
+    if (!currentUser.isEqual(user)) {
+      currentUser = user;
+      client.asyncQueue.enqueueRetryable(async () => {
+        await handleUserChange(offlineComponentProvider.localStore, user);
```

Comment:
I moved the user change guard for LocalStore's `handleUserChange` here. It has always been part of the SyncEngine's call into LocalStore (https://github.com/firebase/firebase-js-sdk/blob/master/packages/firestore/src/core/sync_engine.ts#L1084). This part of the code handles the user change when SyncEngine is not yet loaded (for a fully offline client).

I do however think that the guard here doesn't make a huge difference and mostly added it to be technically correct. This line of code is only executed when memory persistence is used without network access. The only sensible use case for this is an offline Firestore experience that is backed by Bundles. Bundles don't contain any mutations, so the user change would always be a no-op here. Once the user adds the first mutation, SyncEngine gets loaded and overwrites this listener.

I probably didn't describe this very well. But at least I tried, right? :)

Reply:
It's not unreasonable to have here. The change is queued vs being immediately applied and also provides a tiny bit more predictability to the system.